man/out71.1
.TH out71 1 05-November-2024 "LIF Utilities" "LIF Utilities"
.SH NAME
out71 \- a filter to send a program to an HP-71B
.SH SYNOPSIS
.B lifutils out71 [\-o] 
.I FILETYPE
<
.I Input_file
> 
.I Output_device
.PP
.B out71 \-?
.SH DESCRIPTION
The HP-71B Handheld Computer fitted with the HP82401A HPIL Interface 
module can read files not only from a mass storage device but also from 
any other HPIL device. In the latter case, the HP-71B expects to receive a 
32-byte header, which is identical to the file's directory entry on a LIF 
volume, followed by the file data padded (by NULLs in the case of 
.B out71
) to a multiple of 256 bytes. This program always sends a file called
.I PCFILE
of the user-specified file type.
.PP
.B out71
adds the header and padding and can therefore be used to translate a 
Linux file for direct sending to an HP-71B, for example, via an HP82164 RS232
interface. 
.SH OPTIONS
.TP
.I \-o
Some HP-71 files have their length (in their directory entry, and thus in 
the header added by 
.B out71
 given in nibbles. 
.B out71
assumes that all the nibbles in a Linux file are part of the HP-71B file 
and thus always sets this nibble length to an even number. This can 
cause problems if the original HP-71B file had an odd length in nibbles.
The 
.I \-o
option decreases the nibble count in the header by one, effectively 
truncating the last nibble of the Linux file.
.TP
.I \-?
Print a message giving program usage to standard error and exit.
.TP 
.I FILETYPE
Specify the type of the HP-71 file to be created. A complete list of 
possible file types can be found in 
.B lif_create_entry.c,
but the useful ones for this program include
.I TEXT, BASIC71, LEX71 
, and 
.I SDATA.
This parameter is case-insensitive.
.SH FILES
.I hardware/*
: Information on connecting an HP-71B to a PC
.SH EXAMPLES
If an HP-71B is connected to the Linux device
.I /dev/cua0
via a suitable interface and 
.I mytext
is an HP text format file, then 
.PP
.B lifutils out71 TEXT < mytext >/dev/cua0
.PP
will send the file to the HP-71B, which should be set to receive it, e.g. 
by the command
.B COPY :RS232 TO MYTEXT
.SH REFERENCES
The data expected by an HP-71 when reading a file from a non-mass-storage 
device is documented in the HP-71 HPIL Internal Design Specifications 
Volume 1.
.SH BUGS
The
.I \-o
option is not elegant, but there is no automatic way to decide when an HP-71
file has an odd length in nibbles.
.SH SEE ALSO
.B in71(1)
.SH AUTHOR
.B out71
was written by Tony Duell, ard@p850ug1.demon.co.uk and has been placed 
under the GNU Public License version 2.0.


NEWS
change log information is in debian/changelog


src/lib/out71.c
/* out71.c -- A filter to add a directory entry to the start of a file
   (and pad the file so that it consists of 256 byte blocks) for sending
   directly to an HP71B via a suitable interface */
/* 2000 A. R. Duell, and placed under the GPL */

/* The HP71 can copy files from any HPIL device, not just mass-storage 
   devices. The data it expects to receive from a non-mass-storage device
   (like an interface) consists of :
   
   A 32 byte directory entry. This is identical to the 32 byte entry
   stored in the directory of a LIF volume, except that the starting
   block is set to 0

   The file data, padded (by this program with nulls) to a multiple of
   256 bytes.

   This program creates a suitable directory entry for the file read 
   from standard input  , and then sends it, together with the file data
   to standard output. 

   Further details can be found in the HP71 HPIL IDS */

/* One other problem is that some HP71 file types have their length in 
   nybbles stored in the type-specific bytes of the directory entry.
   Since PC files are stored in bytes, the number of nybbles in a PC
   file is always an even number. In the case of an HP71 file stored on 
   the PC  that originally contained an odd number of nybbles, this 
   program will normally send a file that appears to contain one more
   nybble -- the second half of the last byte of the PC file.

   If the -o command line flag is given, however, this program
   decreases the nybble count in the generated directory entry by
   1 to correct for this problem.

   Yes, it's a kludge, but there's no other way to do it that works
   for arbitrary PC files. */

#include <stdio.h>
#include <stdlib.h>
#include <fcntl.h>
#include "config.h"
#include "lifutils.h"
#include "lif_create_entry.h"
#include "lif_const.h"


int out71_read_file(unsigned char * buffer)
/* Read in the file from standard input */
  {
    int byte_counter; /* index of bytes read */
    int clear_counter; /* counter used for clearing one more block */
    int c; /* next character from standard input */
    byte_counter=0;
    /* Read in the file data */
    while(((c=getchar())!=EOF) && (byte_counter<MAX_LENGTH))
      {
        buffer[byte_counter]=c;
        byte_counter++;
      } 
    /* Is the file too large ? */
    if(byte_counter>=MAX_LENGTH)
      {
        fprintf(stderr,"Input file too large for an HP71\n");
        return(RETURN_ERROR);
      }
    /* Clear one more block of the buffer */
    for(clear_counter=0; clear_counter<BLOCK_SIZE; clear_counter++)
      {
        buffer[byte_counter+clear_counter]=0;
      }
    return(byte_counter); /* return number of bytes read */
  }

void out71_usage(void)
  {
    fprintf(stderr,"Usage : lifutils out71 [-o] FILETYPE < input file > output file\n");
    fprintf(stderr,"        Send a program of type FILETYPE to a HP-71B\n");
    fprintf(stderr,"        -o set this flag if the program length is an odd number of nibbles\n");
    fprintf(stderr,"\n");
  }

int out71(int argc, char **argv)
  {
    /* buffer to hold file */
    unsigned char buffer[MAX_LENGTH+BLOCK_SIZE];
    int filetype; /* 2 byte file type code */
    int bytes_read; /* Size of input file */
    int blocks; /* Number of data blocks in file */
    unsigned char dir_entry[ENTRY_SIZE]; /* Directory entry */
    int odd_flag; /* Set if HP71 file contains odd number of nybbles */
    int option;

    SETMODE_STDIN_BINARY;
    SETMODE_STDOUT_BINARY;

    odd_flag=0;
    /* decode command line flags */
    optind=1;
    while((option=getopt(argc,argv,"o?"))!=-1)
      {
        switch(option)
          {
            case 'o' : odd_flag=1;
                       break;
            case '?' : out71_usage();
                       return(RETURN_OK);
          }
      }
    /* decode the file type string given on the command line */
    if(optind != argc-1)
      {
        fprintf(stderr,"out71 : Must give the file type\n");
        out71_usage();
        return(RETURN_ERROR);
      }
    filetype=get_filetype(argv[optind]);
    if(filetype==-1)
      {
        fprintf(stderr,"out71 : Unknown file type : %s\n",argv[optind]);
        out71_usage();
        return(RETURN_ERROR);
      }

    /* Read in input file */;
    bytes_read=out71_read_file(buffer);
    return_if_error(bytes_read);

    /* Create directory entry */
    create_entry(dir_entry,"PCFILE",filetype,0,bytes_read, odd_flag);
    /* and output it */
    fwrite(dir_entry,sizeof(unsigned char), ENTRY_SIZE, stdout);
    
    /* Find number of blocks to output */
    blocks=(bytes_read+(BLOCK_SIZE-1))/BLOCK_SIZE;
    /* and output them */
    fwrite(buffer,sizeof(unsigned char),blocks*BLOCK_SIZE,stdout);
    return(RETURN_OK);
  }

